pennylane/devices/qubit_mixed/apply_operation.py

```diff
@@ -0,0 +1,600 @@
+            math.conj(k),
+            axes_right,
+        )
+
+    def _tensordot_single_kraus(kraus, state=state):
+        if len(kraus) == 1:
+            _state = _conjugate_state_with(kraus[0], state)
+
+        else:
+            _state = math.sum(math.stack([_conjugate_state_with(k, state) for k in kraus]), axis=0)
+
+        source_left = list(range(num_ch_wires))
+        dest_left = row_wires_list
+        source_right = list(range(-num_ch_wires, 0))
+        dest_right = col_wires_list
+
+        result = math.moveaxis(_state, source_left + source_right, dest_left + dest_right)
+
+        return math.cast_like(result, state)
+
+    if is_op_channel or not is_mat_batched:  # First deal with the channel case
+        return _tensordot_single_kraus(kraus)
+    # Due to the limit of tensordot we better deal with each batch separately
+    kraus_batch = [[k[batch_i] for k in kraus] for batch_i in range(batch_size)]
+    if not is_state_batched:
+        return math.stack([_tensordot_single_kraus(kraus_batch_i) for kraus_batch_i in kraus_batch])
+    # If the both mat and state are batched, we can directly apply the tensordot
+    row_wires_list = channel_wires.tolist()
+    col_wires_list = [w + num_wires for w in row_wires_list]
+    axes_left = [channel_col_ids, row_wires_list]
+    axes_right = [col_wires_list, channel_col_ids]
+    kraus = math.reshape(kraus, state.shape)
+    return math.stack(
+        [_tensordot_single_kraus([kraus[batch_i]], state[batch_i]) for batch_i in range(batch_size)]
+    )
+
+
+@singledispatch
+def apply_operation(
+    op: qml.operation.Operator,
+    state,
+    is_state_batched: bool = False,
+    debugger=None,
+    **_,
+):
+    """Apply an operation to a given state.
+
+    Args:
+        op (Operator): The operation to apply to ``state``
+        state (TensorLike): The starting state.
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+        debugger (_Debugger): The debugger to use
+
+    Keyword Arguments:
+        rng (Optional[numpy.random._generator.Generator]): A NumPy random number generator.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. Only for simulation using JAX.
+            If None, a ``numpy.random.default_rng`` will be used for sampling.
+        tape_shots (Shots): the shots object of the tape
+
+    Returns:
+        ndarray: output state
+
+    .. warning::
+
+        ``apply_operation`` is an internal function, and thus subject to change without a deprecation cycle.
+
+    .. warning::
+        ``apply_operation`` applies no validation to its inputs.
+
+        This function assumes that the wires of the operator correspond to indices
+        of the state. See :func:`~.map_wires` to convert operations to integer wire labels.
+
+        The shape of state should be ``[QUDIT_DIM]*(num_wires * 2)``, where ``QUDIT_DIM`` is
```

Comment:
Can you explain the structure of such a state? The shape of a density matrix is $d^n\times d^n$ where $d$ is the dimension of the system and $n$ is the number of wires. How does a density matrix map to the state that you are using?